Libertydevops openliberty microprofile configuration (#2469)
OpenLiberty/blogs: posts/2022-08-15-openliberty-microprofile-configuration.adoc

layout: post
title: "Open Liberty and MicroProfile configuration file management in multiple Kubernetes environments"
categories: blog
author_picture: https://avatars3.githubusercontent.com/shamjithantholi
author_github: https://github.com/shamjithantholi
seo-title: Open Liberty and MicroProfile configuration file management in multiple Kubernetes environments
seo-description: Methods to manage Open Liberty and MicroProfile configuration files in multiple Kubernetes environments.
blog_description: Methods to manage Open Liberty and MicroProfile configuration files in multiple Kubernetes environments.
open-graph-image: https://openliberty.io/img/blog/liberty-devops-generic-architecture.png
---
= Open Liberty and MicroProfile configuration file management in multiple Kubernetes environments
Shamjith Antholi <https://github.com/shamjithantholi>
:imagesdir: /
:url-prefix:
:url-about: /

[#Intro]

== What is MicroProfile and Open Liberty

The link:https://microprofile.io[MicroProfile] specification consists of a collection of Enterprise Java APIs and technologies that together form a core baseline for microservices that aims to deliver application portability across multiple runtimes.

> link:https://openliberty.io/?_ga=2.75178834.1545551050.1658778682-1210051418.1658778682[Open Liberty] is a lightweight open source Java™ runtime that is built by using modular features. link:https://www.ibm.com/cloud/websphere-liberty[WebSphere Liberty] is a commercial version of Open Liberty.

Reviewer: I don't think there is any need to mention WebSphere Liberty

Reviewer: What is that query parameter for? `?_ga=2.75178834.1545551050.1658778682-1210051418.1658778682`

Author: Removed WebSphere Liberty reference and query parameter

== About MicroProfile config and containerization

> Externalization of constants into different file formats like .java, .properties, .xml etc are a common coding practice to separete application code from the configuration variables. Variables in the source code are resolved from these 'constants' files on run time. In this blog, we will be discussing how Open Liberty and MicroProfile manages such externalization requirements.  

Reviewer: `.java` isn't a file format that you would use to externalize constants/configuration. Even if you move your constants to a single Java class, they're still internal, not external.

Reviewer: Also, we're talking about configuration here, not constants ... if something is truely constant then there is no need to externalize it at all.

```suggestion
Externalizing configuration into different file formats like JSON, XML and Java Properties files are a common coding practice to separate application code from the configuration variables. Variables in the source code are initialized from these configuration files at runtime. In this blog, we will be discussing how Open Liberty and MicroProfile manage such configuration requirements.  
```

Author: Done

> MicroProfile and Open Liberty have standardized the use of these parameter files with additional properties like ordianls. Different type of files where the constants and run time variables can be hosted are 

Reviewer: You probably need to introduce the idea that there are lots of possible config "sources" (not just files).
```suggestion
MicroProfile and Open Liberty have standardized the use of these configuration sources. Some commonly used sources are;
```

Author: Done

             server.xml

>              Environment based files/variables in source code

Reviewer: ```suggestion
             Environment Variables
             System Properties
```

Author: Done

             microprofile-config.properties

>              bootstrap.properties

Reviewer: Use caution with this one because the values in a bootstrap.properties files just end up loaded as System Properties at runtime.

Author: Removed bootstrap.properties

>              CustomConfig properties (anything which implement org.eclipse.microprofile.config.spi.ConfigSource)

Reviewer: ```suggestion
             Other custom config sources (implemented using org.eclipse.microprofile.config.spi.ConfigSource)
```

Author: Done

> All these property files contains an ordinal value to set the reading order of the file by the application. 

Reviewer: You might want to link to a Config guide or blog post here for more details?
```suggestion
These config sources may contain a special ordinal value to control the order of precedence at runtime.
```

Author: Done

> Applications need to be hosted on multiple environments like development, test and production. There will be several variables whose values would be changing across these environments and containerized platforms has different configuration methods like given below to inject the variable values on run time to the application.

Reviewer: ```suggestion
In many cases, applications need to be hosted in multiple environments like development, test and production. Some of the configuration may need to set differently depending on the environment and platform, whereas some may remain the same. In a containerized environment, there are number of different configuration methods, such as;
```

Author: Done

Architecture diagram of Open Liberty config files flow from GitHub through a build tool to Kubernetes is as given below

image::/img/blog/ol-configuration1.png[Open Liberty configuration file flow,width=100%,align="left"]

== Different data injection methods in Kubernetes/OpenShift

There are multiple ways to inject data into application containers. This blog will discuss those methods one by one. The commands given in this blogs are OpenShift commands, but you can run the same on Kubernetes by replacing 'oc' by 'kubeclt'

For the purpose of demonstration, i have defined variables in below given files and the values for the same is injected from different container data sources.

server.xml&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;microprofile-config.properties

Reviewer: This can't be the best way to format these headings

Author: Will check better option later
